kernel/src/main/java/org/kframework/kast/KastFrontEnd.java

```diff
@@ -80,26 +94,89 @@ public static List<com.google.inject.Module> getModules() {
+                    Tuple2<Either<Set<KEMException>, K>, Set<KEMException>> res = parseInModule.parseString(stringToParse, sort, source);
+                    kem.addAllKException(res._2().stream().map(KEMException::getKException).collect(Collectors.toSet()));
+                    if (res._1().isLeft()) {
+                        throw res._1().left().get().iterator().next();
+                    }
+                    // important to get the extension module for unparsing because it contains generated syntax
+                    // like casts, projections and others
+                    Module unparsingMod = parseInModule.getExtensionModule();
+                    K parsed = new TreeNodesToKORE(Outer::parseSort, true).down(res._1().right().get());
+
+                    if (options.expandMacros) {
+                        parsed = ExpandMacros.forNonSentences(unparsingMod, files.get(), def.kompileOptions, false).expand(parsed);
+                    }
+                    ConfigurationInfoFromModule configInfo = new ConfigurationInfoFromModule(mod);
+                    LabelInfo labelInfo = new LabelInfoFromModule(mod);
+                    SortInfo sortInfo = SortInfo.fromModule(mod);
+
+                    Rule r = Rule.apply(parsed, BooleanUtils.TRUE, BooleanUtils.TRUE, Att.empty());
+                    if (options.steps.contains(KompileSteps.macros))
+                        r = (Rule) ExpandMacros.forNonSentences(unparsingMod, files.get(), def.kompileOptions, false).expand(r);
+                    if (options.steps.contains(KompileSteps.closeCells))
+                        r = (Rule) new CloseCells(configInfo, sortInfo, labelInfo).close(r);
+                    if (options.steps.contains(KompileSteps.resolveCasts)) // move casts to side condition predicates
+                        r = (Rule) new ResolveSemanticCasts(false).resolve(r);
+                    if (options.steps.contains(KompileSteps.addImplicitComputationCell))
+                        r = (Rule) new AddImplicitComputationCell(configInfo, labelInfo).apply(mod, r);
+                    if (options.steps.contains(KompileSteps.concretizeCells))
+                        r = (Rule) new ConcretizeCells(configInfo, labelInfo, sortInfo, mod, true).concretize(mod, r);
+                    K result = options.steps.contains(KompileSteps.body) ? r.body() : KApply(KLabel("#ruleRequiresEnsures"), KList(r.body(), r.requires(), r.ensures()));
+                    kprint.get().prettyPrint(def.kompiledDefinition, unparsingMod, s -> kprint.get().outputFile(s), result, sort);
                 }
+
+                sw.printTotal("Total");
+                return 0;
+            } else if (!options.steps.equals(Lists.newArrayList(KastFrontEnd.KompileSteps.closeCells, KastFrontEnd.KompileSteps.resolveCasts, KastFrontEnd.KompileSteps.body))) {
+                throw KEMException.innerParserError("Option --steps is available only with --input rule.");
             }
 
```

Comment:
This seems like a less good way to tell that the option wasn't set. Can we just store a Null value if it's unset, and then if it's Null set the default here? It seems like a bad idea to have it be that we can't tell between the option not being set, and the option being set to the default (the user could say `--steps closeCells,resolveCasts,body` on the CLI, without `--input rule`, and this check wouldn't trigger).

Reply:
I did it this way because I would like to have `kast -h` say what the default steps are:
```
--steps
      Apply specified kompilation steps to the parsed term. Only for --input 
      rule. Use --steps help for a detailed description of available steps.
      Default: [closeCells, resolveCasts, body]
```
Setting `null` is going to say `Default: []`
I'm not sure if there is a way to actually check if the option has actually been provided.
In any case, it's ignored if `--input` is not `rule`